File: py-package/src/tlang_read_node/read_node.py
from __future__ import annotations

import json
import pickle
import re
from pathlib import Path
from typing import Any, BinaryIO, Callable


FIXTURE_LOGS = {"build_log_ocaml_mock.json", "build_log_legacy_version.json"}


def _validate_non_empty_string(value: str, arg: str) -> None:
    if not isinstance(value, str) or not value.strip():
        raise ValueError(f"`{arg}` must be a non-empty string.")


def _pipeline_path(pipeline_dir: str | Path) -> Path:
    path = Path(pipeline_dir)
    if not str(path).strip():
        raise ValueError("`pipeline_dir` must be a non-empty path.")
    return path


def _should_filter_fixture_logs(pipeline_dir: Path) -> bool:
    repo_root = pipeline_dir.parent
    return (repo_root / "src" / "pipeline" / "builder_logs.ml").is_file()


def _list_build_logs(pipeline_dir: Path) -> list[str]:
    logs = sorted(
        [path.name for path in pipeline_dir.glob("build_log_*.json")],
        reverse=True,
    )

    if (
        _should_filter_fixture_logs(pipeline_dir)
        and len(logs) > 1
        and any(log not in FIXTURE_LOGS for log in logs)
    ):
        logs = [log for log in logs if log not in FIXTURE_LOGS]

    return logs


def _select_build_log(logs: list[str], which_log: str | None, pipeline_dir: Path) -> str:
    if which_log is None:
        if not logs:
            raise FileNotFoundError(
                f"No build logs found in `{pipeline_dir}`. Build the pipeline first."
            )
        return logs[0]

    _validate_non_empty_string(which_log, "which_log")

    try:
        pattern = re.compile(which_log)
    except re.error as err:
        raise ValueError(f"Invalid regular expression for `which_log`: {err}") from err

    matches = [log for log in logs if pattern.search(log)]
    if not matches:
        raise FileNotFoundError(
            f'No build logs found in `{pipeline_dir}` matching "{which_log}".'
        )

    return matches[0]


def _read_build_log(log_path: Path) -> dict[str, Any]:
    try:
        with log_path.open("r", encoding="utf-8") as handle:
            data = json.load(handle)
    except OSError as err:
        raise OSError(f"Failed to read build log `{log_path.name}`: {err}") from err
    except json.JSONDecodeError as err:
        raise ValueError(f"Failed to read build log `{log_path.name}`: {err}") from err

    if not isinstance(data, dict):
        raise ValueError(f"Build log `{log_path.name}` must decode to an object.")

    return data


def _find_node_entry(nodes: Any, name: str, log_file: str) -> dict[str, Any]:
    if not isinstance(nodes, list):
        raise ValueError(f"Build log `{log_file}` does not contain a `nodes` array.")

    for entry in nodes:
        if isinstance(entry, dict) and entry.get("node") == name:
            return entry

    raise KeyError(f"Node `{name}` not found in build log `{log_file}`.")


def _resolve_artifact_path(path_value: Any, pipeline_dir: Path) -> Path:
    if not isinstance(path_value, str) or not path_value.strip():
        raise ValueError("Node entry does not contain a valid artifact path.")

    artifact_path = Path(path_value)
    if artifact_path.is_absolute():
        return artifact_path

    return (pipeline_dir.parent / artifact_path).resolve(strict=False)


def deserialize(path: str | Path) -> Any:
    artifact_path = Path(path)
    loaders: list[tuple[str, Callable[[BinaryIO], Any]]] = [("pickle", pickle.load)]

    try:
        import dill  # type: ignore
    except ImportError:
        pass
    else:
        loaders.append(("dill", dill.load))

    try:
        import cloudpickle  # type: ignore
    except ImportError:
        pass
    else:
        loaders.append(("cloudpickle", cloudpickle.load))

    errors: list[str] = []
    for loader_name, loader in loaders:
        try:
            with artifact_path.open("rb") as handle:
                return loader(handle)
        except Exception as err:  # noqa: BLE001
            errors.append(f"{loader_name}: {type(err).__name__}: {err}")

    raise RuntimeError(
        f"Failed to deserialize `{artifact_path}`. Attempted loaders: "
        + "; ".join(errors)
    )


def read_node(
    name: str,
    which_log: str | None = None,
    pipeline_dir: str | Path = "_pipeline",
    deserializer: Callable[[str | Path], Any] = deserialize,
) -> Any:
    """Read a node artifact from a built T pipeline.

    When ``which_log`` is ``None``, the helper picks the first
    reverse-alphabetically sorted ``build_log_*.json`` file, which matches T's
    timestamped log naming and therefore resolves to the most recent build.
    """

    _validate_non_empty_string(name, "name")
    pipeline_path = _pipeline_path(pipeline_dir)

    if not pipeline_path.is_dir():
        raise FileNotFoundError(f"Pipeline directory `{pipeline_path}` does not exist.")

    if not callable(deserializer):
        raise TypeError("`deserializer` must be callable.")

    logs = _list_build_logs(pipeline_path)
    log_file = _select_build_log(logs, which_log, pipeline_path)
    build_log = _read_build_log(pipeline_path / log_file)
    node_entry = _find_node_entry(build_log.get("nodes"), name, log_file)
    artifact_path = _resolve_artifact_path(node_entry.get("path"), pipeline_path)

    try:
        return deserializer(artifact_path)
    except Exception as err:  # noqa: BLE001
        raise RuntimeError(
            f"Failed to deserialize node `{name}` from `{artifact_path}`: {err}"
        ) from err

File: r-package/R/read_node.R
#' Validate that a value is a scalar, non-empty, and non-NA character string
#'
#' @param x The value to check.
#' @param arg Character. The argument name to use in the error message.
#'
#' @return None. Throws an error if the validation fails.
#'
#' @keywords internal
validate_scalar_string <- function(x, arg) {
  if (!is.character(x) || length(x) != 1L || is.na(x) || !nzchar(x)) {
    stop(sprintf("`%s` must be a non-empty string.", arg), call. = FALSE)
  }
}

#' List build logs in the pipeline directory, sorted reverse-alphabetically
#'
#' Searches the specified pipeline directory for build log JSON files, sorts them,
#' and filters out fixture logs if this is executed within the project repository checkout.
#'
#' @param pipeline_dir Character. The path to the pipeline directory.
#'
#' @return A sorted character vector of build log filenames.
#'
#' @keywords internal
list_build_logs <- function(pipeline_dir) {
  logs <- list.files(
    pipeline_dir,
    pattern = "^build_log_.*\\.json$",
    full.names = FALSE
  )
  logs <- sort(logs, decreasing = TRUE)

  builder_logs_ml_path <- file.path(dirname(pipeline_dir), "src", "pipeline", "builder_logs.ml")

  # Mirror T's fixture-log filtering only when reading from a repository
  # checkout, where these internal fixture logs can live beside real build logs.
  fixture_logs <- c("build_log_ocaml_mock.json", "build_log_legacy_version.json")
  if (file.exists(builder_logs_ml_path) && length(logs) > 1L && any(!logs %in% fixture_logs)) {
    logs <- logs[!logs %in% fixture_logs]
  }

  logs
}

#' Select a build log from a list of logs
#'
#' Selects the latest log file if `which_log` is `NULL`, or selects the first
#' log file matching the regular expression `which_log`.
#'
#' @param logs Character vector of available build log filenames.
#' @param which_log Character or NULL. A regular expression pattern to select a build log.
#' @param pipeline_dir Character. The path to the pipeline directory (used in error messages).
#'
#' @return Character. The selected build log filename.
#'
#' @keywords internal
select_build_log <- function(logs, which_log, pipeline_dir) {
  if (is.null(which_log)) {
    if (length(logs) == 0L) {
      stop(
        sprintf(
          "No build logs found in `%s`. Build the pipeline first.",
          pipeline_dir
        ),
        call. = FALSE
      )
    }

    return(logs[[1L]])
  }

  validate_scalar_string(which_log, "which_log")

  matches <- tryCatch(
    grep(which_log, logs, value = TRUE),
    error = function(err) {
      stop(
        sprintf(
          "Invalid regular expression for `which_log`: %s",
          conditionMessage(err)
        ),
        call. = FALSE
      )
    }
  )

  if (length(matches) == 0L) {
    stop(
      sprintf(
        "No build logs found in `%s` matching \"%s\".",
        pipeline_dir,
        which_log
      ),
      call. = FALSE
    )
  }

  matches[[1L]]
}

#' Read and parse a JSON build log
#'
#' Parses a single build log JSON file into an R list structure.
#'
#' @param log_path Character. The path to the build log file.
#'
#' @return A list containing the parsed JSON contents.
#'
#' @keywords internal
read_build_log <- function(log_path) {
  tryCatch(
    jsonlite::fromJSON(log_path, simplifyVector = FALSE),
    error = function(err) {
      stop(
        sprintf(
          "Failed to read build log `%s`: %s",
          basename(log_path),
          conditionMessage(err)
        ),
        call. = FALSE
      )
    }
  )
}

#' Find a node's entry in a build log's nodes array
#'
#' Iterates through the list of nodes from a build log and returns the entry
#' matching the specified name.
#'
#' @param nodes A list of node entries from the build log.
#' @param name Character. The name of the node to find.
#' @param log_file Character. The filename of the log (used in error messages).
#'
#' @return A list representing the matched node entry.
#'
#' @keywords internal
find_node_entry <- function(nodes, name, log_file) {
  if (!is.list(nodes)) {
    stop(
      sprintf("Build log `%s` does not contain a `nodes` array.", log_file),
      call. = FALSE
    )
  }

  for (entry in nodes) {
    if (is.list(entry) && identical(entry$node, name)) {
      return(entry)
    }
  }

  stop(
    sprintf("Node `%s` not found in build log `%s`.", name, log_file),
    call. = FALSE
  )
}

#' Resolve an artifact path to an absolute path
#'
#' Converts a relative artifact path from the build log into an absolute path,
#' supporting both Unix and Windows absolute path styles.
#'
#' @param path Character. The raw artifact path.
#' @param pipeline_dir Character. The path to the pipeline directory.
#'
#' @return Character. The resolved absolute path.
#'
#' @keywords internal
resolve_artifact_path <- function(path, pipeline_dir) {
  if (!is.character(path) || length(path) != 1L || is.na(path) || !nzchar(path)) {
    stop("Node entry does not contain a valid artifact path.", call. = FALSE)
  }

  # Match Unix absolute paths (`/path/...`) and standard Windows drive-letter
  # absolute paths (`C:/path/...` or `C:\\path\\...`).
  if (grepl("^(/|[A-Za-z]:[/\\\\])", path)) {
    return(path)
  }

  normalizePath(
    file.path(dirname(pipeline_dir), path),
    winslash = "/",
    mustWork = FALSE
  )
}

#' Read a node artifact from a built T pipeline
#'
#' Reads the latest build log (or a matching historical one) from the
#' `_pipeline/` directory, locates the requested node, and deserializes the
#' node artifact. When `which_log` is `NULL`, the helper picks the first
#' reverse-alphabetically sorted `build_log_*.json` file, which matches T's
#' timestamped log naming and therefore resolves to the most recent build.
#'
#' @param name Name of the node to read.
#' @param which_log Optional regular expression used to select a specific build
#'   log filename. Defaults to the latest available build log.
#' @param pipeline_dir Path to the pipeline build directory. Defaults to
#'   `"_pipeline"`.
#' @param deserializer Function used to deserialize the artifact file. Defaults
#'   to `readRDS()`.
#' @param return_path Logical. If `TRUE`, returns the path to the artifact
#'   instead of deserializing it. Defaults to `FALSE`.
#'
#' @return The deserialized node artifact, or the path to it if `return_path` is `TRUE`.
#'
#' @details
#' The function locates the pipeline folder and reads the selected build log. If
#' `return_path` is set to `TRUE`, it returns the absolute system path to the
#' serialized artifact instead of deserializing the object. Otherwise, it utilizes
#' the `deserializer` function to deserialize the R object.
#'
#' @examples
#' \init{
#'   # Assuming node "clean_data" was built successfully
#'   df <- read_node("clean_data")
#' }
#'
#' @export
read_node <- function(
    name,
    which_log = NULL,
    pipeline_dir = "_pipeline",
    deserializer = readRDS,
    return_path = FALSE) {
  validate_scalar_string(name, "name")
  validate_scalar_string(pipeline_dir, "pipeline_dir")

  if (!dir.exists(pipeline_dir)) {
    stop(
      sprintf("Pipeline directory `%s` does not exist.", pipeline_dir),
      call. = FALSE
    )
  }

  if (!is.function(deserializer)) {
    stop("`deserializer` must be a function.", call. = FALSE)
  }

  logs <- list_build_logs(pipeline_dir)
  log_file <- select_build_log(logs, which_log, pipeline_dir)
  log_path <- file.path(pipeline_dir, log_file)
  build_log <- read_build_log(log_path)
  node_entry <- find_node_entry(build_log$nodes, name, log_file)
  artifact_path <- resolve_artifact_path(node_entry$path, pipeline_dir)

  if (isTRUE(return_path)) {
    return(artifact_path)
  }

  tryCatch(
    deserializer(artifact_path),
    error = function(err) {
      stop(
        sprintf(
          "Failed to deserialize node `%s` from `%s`: %s",
          name,
          artifact_path,
          conditionMessage(err)
        ),
        call. = FALSE
      )
    }
  )
}
